tests/api_internal/test_internal_api_call.py

```diff
@@ -0,0 +1,151 @@
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+
+from __future__ import annotations
+
+import json
+import unittest
+from unittest import mock
+
+import requests
+
+from airflow.api_internal.internal_api_call import InternalApiConfig, internal_api_call
+from airflow.serialization.serialized_objects import BaseSerialization
+from tests.test_utils.config import conf_vars
+
+
+class TestInternalApiConfig(unittest.TestCase):
+    def setUp(self):
+        InternalApiConfig._initialized = False
+
+    @conf_vars(
+        {
+            ("core", "database_access_isolation"): "false",
+            ("core", "database_api_url"): "http://localhost:8888",
+        }
+    )
+    def test_get_use_internal_api_disabled(self):
+        self.assertFalse(InternalApiConfig.get_use_internal_api())
+
+    @conf_vars(
+        {
+            ("core", "database_access_isolation"): "true",
+            ("core", "database_api_url"): "http://localhost:8888",
+        }
+    )
+    def test_get_use_internal_api_enabled(self):
+        self.assertTrue(InternalApiConfig.get_use_internal_api())
+        self.assertEqual(
+            InternalApiConfig.get_internal_api_endpoint(),
+            "http://localhost:8888/internal/v1/rpcapi",
+        )
+
+
+@internal_api_call
+def fake_method() -> str:
+    return "local-call"
+
+
+@internal_api_call
+def fake_method_with_params(dag_id: str, task_id: int) -> str:
+    return f"local-call-with-params-{dag_id}-{task_id}"
+
+
+class TestInternalApiCall(unittest.TestCase):
+    def setUp(self):
+        InternalApiConfig._initialized = False
+
+    @conf_vars(
+        {
+            ("core", "database_access_isolation"): "false",
+            ("core", "database_api_url"): "http://localhost:8888",
+        }
+    )
+    @mock.patch("airflow.api_internal.internal_api_call.requests")
+    def test_local_call(self, mock_requests):
+        result = fake_method()
+
+        self.assertEqual(result, "local-call")
+        mock_requests.post.assert_not_called()
+
+    @conf_vars(
+        {
+            ("core", "database_access_isolation"): "true",
+            ("core", "database_api_url"): "http://localhost:8888",
+        }
+    )
+    @mock.patch("airflow.api_internal.internal_api_call.requests")
+    def test_remote_call(self, mock_requests):
+        response = requests.Response()
+        response.status_code = 200
+
+        response._content = json.dumps(BaseSerialization.serialize("remote-call"))
+
+        mock_requests.post.return_value = response
+
+        result = fake_method()
+        self.assertEqual(result, "remote-call")
+        expected_data = json.dumps(
+            {
+                "jsonrpc": "2.0",
+                "method": "tests.api_internal.test_internal_api_call.fake_method",
+                "params": '{"__var": {}, "__type": "dict"}',
```

Comment:
Where does it come from?

Reply:
method  -> fake_method in this file
params -> this is what is send for empty params (this is what BaseSerializator does)

Reply:
In that case, I would maybe do it that way to makes more sense. The concern I have is, probably other folks will wonder where these `__var` and `__type` come from.

```
"params": BaseSerialization.serialize({})
```